src/provider/mod.rs

```diff
@@ -422,54 +582,170 @@ where
                 L1CommunicationError::Custom("Error occurred while estimating L1 base fees.")
             })?;
         let max_fee_per_gas = base_l1_fees_data.max_fee_per_gas + max_priority_fee_per_gas;
-        //
 
-        let l2_base_cost = bridge_hub_contract
-            .l2TransactionBaseCost(
-                U256::from(l2_chain_id),
-                U256::from(max_fee_per_gas),
-                estimate_gas_l1_to_l2,
-                gas_per_pubdata_limit,
-            )
-            .call()
+        let bridge_hub_contract_address = self
+            .get_bridgehub_contract()
             .await
             .map_err(|_| {
                 L1CommunicationError::Custom(
-                    "Error occurred while estimating L2 transaction base cost.",
+                    "Error occurred while fetching the bridge hub contract address.",
                 )
             })?
-            ._0;
+            .unwrap();
+        let bridge_hub_contract = Bridgehub::new(bridge_hub_contract_address, l1_provider);
 
-        let l1_value = l2_base_cost + amount;
+        if deposit_request.token != ETHER_L1_ADDRESS {
```

Comment:
Given that we're about to get a few more options here (deposit custom token, deposit ERC20 to a custom base token chain), probably it makes sense to extract this logic into a separate function. The function probably shouldn't live on the `ZksyncProviderWithWallet` trait (to not create too many public methods here).

The way [it's currently implemented in zksync-ethers](https://github.com/zksync-sdk/zksync-ethers/blob/main/src/adapters.ts#L416) makes sense historically (as the number of supported combinations has grown) but doesn't feel maintainable, so I think we need different abstractions here.

We have several logical steps here:
1. Approve required tokens if needed. We may have approve either 0, 1, or 2 tokens:
  - Deposit ETH to ETH chain -> 0 tokens
  - Deposit ERC20 to ETH chain -> 1 token
  - Deposit native ERC20 to ERC20 chain -> 1 token
  - Deposit non-native ERC20 to ERC20 chain -> 2 tokens
2. Prepare tx data for L1. There are two options here:
  - `requestL2TransactionDirect` for native tokens
  - `requestL2TransactionTwoBridges` for non-native tokens (it also includes depositing ETH to non-ETH based token)
3. Set remaining fields (like `fee_per_gas`) and submit transaction. This is shared for all the tokens.
4. Get priority op response. This is shared for everyone.

What I'm thinking currently is something like this (pseudocode):

```rust
pub struct DepositExecutor<'a> {
  l1_provider: &'a L1Provider,
  l2_provider: &'a L2Provider,
  request: &'a DepositRequest
}

impl<'a> DepositExecutor<'a> {
  pub async fn execute(self) -> ... {
    let l1_fee_params = self.get_l1_fee_params().await?;
    // Get the final transaction to be submitted. We want to do it before
    // approving to decrease chances of deposit failing _after_ submitting any txs.
    let l1_deposit_tx = self.encode_deposit_calldata(&l1_fee_params).await?;
    // Approve required amount of tokens if needed. Can submit from 0 to 2 transactions.
    self.approve_tokens(&l1_fee_params).await?;
    // Now everything is ready, submit the deposit.
    self.submit(l1_deposit_tx).await
  }
}
```

Then the deposit function will be as simple as:

```rust
async fn deposit(&self, deposit_request: &DepositRequest, l1_provider: &P) -> ... {
  let executor = DepositExecutor { 
    l1_provider,
    l2_provider: self,
    request: deposit_request
  };
  executor.execute().await
}
```

Additional benefit here is that we will be able to later extract parts of it (e.g. provide an interface to just encode deposit calldata or just approve enough tokens)
easily without creating too much copy-paste.

Reply:
I added the `DepositExecutor` abstraction and the provider looks much cleaner. The `execute` method in the `DepositExecutor` is not as lean as in you pseudocode as we need to share some common state among the methods inside (l2 chain id, sender, receiver etc). Please, take a look.